src/main/scala/io/iohk/ethereum/network/PeerStatisticsActor.scala

```diff
@@ -0,0 +1,75 @@
+package io.iohk.ethereum.network
+
+import akka.actor._
+import io.iohk.ethereum.network.PeerEventBusActor._
+import io.iohk.ethereum.network.p2p.messages.Codes
+import scala.concurrent.duration.FiniteDuration
+import cats.kernel.Monoid
+
+class PeerStatisticsActor(
+    peerEventBus: ActorRef,
+    var maybeStats: Option[TimeSlotStats[PeerId, PeerStatisticsActor.Stat]]
+) extends Actor {
+  import PeerStatisticsActor._
+
+  // Subscribe to messages received from handshaked peers to maintain stats.
+  peerEventBus ! Subscribe(MessageSubscriptionClassifier)
+  // Removing peers is an optimisation to free space, but eventualy the stats would be overwritten anyway.
+  peerEventBus ! Subscribe(SubscriptionClassifier.PeerDisconnectedClassifier(PeerSelector.AllPeers))
+
+  def receive: Receive = handlePeerEvents orElse handleStatsRequests
+
+  private def handlePeerEvents: Receive = {
+    case PeerEvent.MessageFromPeer(_, peerId) =>
+      maybeStats = maybeStats.map(_.add(peerId, Stat(responsesReceived = 1)))
+
+    case PeerEvent.PeerDisconnected(peerId) =>
+      maybeStats = maybeStats.map(_.remove(peerId))
+  }
+
+  private def handleStatsRequests: Receive = {
+    case GetStatsForAll =>
+      sender ! StatsForAll(maybeStats.map(_.getAll()).getOrElse(Map.empty))
+
+    case GetStatsForPeer(peerId) =>
+      sender ! StatsForPeer(peerId, maybeStats.map(_.get(peerId)).getOrElse(Stat.empty))
+  }
+}
+
+object PeerStatisticsActor {
+  case class Stat(
+      responsesReceived: Int
```

Comment:
to consider, maybe it is worth having also the timestamp when we heard from the peer last time ?

Reply:
Okay I'll add first and last observation times.

Reply:
Done.